docs(tree): add merge semantics docs (#20841)
microsoft/FluidFramework: packages/dds/tree/docs/main/merge-semantics.md

# SharedTree Merge Semantics

This document offers a high-level description `SharedTree`'s merge semantics.

Target audience: `SharedTree` users and maintainers.

## What Are Merge Semantics?

Merge semantics define how `SharedTree` reconciles concurrent edits.

### Concurrent Edits

When several peers edit the same document, it's possible for some of the edits to be concurrent.
The edits of two clients are concurrent if they each made those edits before they had received the others’ edits from the server.

> For example, imagine Alice and Bob are editing a document that contains stick notes whose background color can be changed.

Reviewer: ```suggestion
For example, imagine Alice and Bob are editing a document that contains sticky notes whose background color can be changed.
```

This is not a case of concurrent editing because Bob had received Alice's edit before changing the color.

On the other hand, if Bob makes this change _before_ receiving the edit from Alice,
then his edit will be concurrent to Alice's edit, and he will be changing the color from yellow to blue.

### Reconciling Concurrent Edits

Reconciling concurrent edits is trivial when they affect independent parts of the tree.

For example, if Alice and Bob concurrently change the background color of _separate_ sticky notes,
then there is only one reasonable outcome.

However, it's possible for some concurrent edits to affect overlapping parts of the tree.
This leads to a situation where there may be multiple reasonable outcomes
(and possibly some unreasonable ones).

For example, if Alice and Bob concurrently change the background color of _the same_ sticky note
such that Alice would change it from yellow to red and Bob would change it from yellow to blue,
then there are multiple reasonable outcomes:

-   change the color to red
-   change the color to blue
-   keep the color yellow

> 

Author: Done!

preconditions can be used to prevent an edit/transaction from being applied
if concurrent edits that were sequenced earlier have impacted the document in a way that violates some specific requirements.

By default, that is, in the absence of [constraints](#constraints),
the only precondition that must be met in order for an edit to apply,
is that the edit must not violate the document's schema.

For example,
when moving a node to a specific destination only requires that the node's type be allowed by the schema of the destination.

One could imagine a host of potential additional preconditions for such an edit:

-   The node being moved must not have been concurrently moved
-   The node being moved must not have been concurrently removed
-   Ancestors of the node being moved must not have been concurrently removed
-   The parent and ancestors of destination must not have been concurrently removed

`SharedTree`'s merge semantics are such that, by default, only the schema matters.
The benefits of not violating the schema are obvious,
but the benefits of not having more preconditions are more subtle.
Indeed, while no one should wish for their edit to violate the schema,

> one may wish for their edit not to be carried out if some other precondition is not met.

Reviewer: "While" gives an expectation that the two clauses are in some kind of opposition, but they don't seem to be.

Author: I rephrased the section

then there would be no way for users to opt-out of these preconditions,
thereby limiting their ability to express a more permissive edit.

In the rest of this section, we call out some interesting implications that follow from the doctrine of minimal preconditions,
the last of which is also a key benefit of this doctrine.

#### Removal Is Movement

`SharedTree` allows subtrees to be removed,
such as when an element from an array node is removed,
a key is deleted from a map node,
or when the field on an object is overwritten.
This removal does not adversely affect concurrent edits that target the removed subtree.

Consider the following scenario:
Alice removes a whole list of sticky notes, while Bob concurrently moves a sticky note out of that list and into another (non-removed) list.
`SharedTree`'s removal semantics ensure that Bob's edit will apply no matter the sequencing order.
If that weren't the case, then there would be a race between Alice and Bob's edit,
where Bob's edit would not apply if Alice's edit were sequenced first.

Note that in a lot of cases, the changes to the removed subtree won't be immediately visible because the tree is removed.

> They will, however, become visible if that removal is undone.

Author: You're not wrong, but the thing is:
1. We don't yet have that doc
2. I do mean undone in the most general sense (for example, the element or one of its ancestors could be moved back into the tree)
3. This is a case where it seems okay to rely on the reader's intuitive understanding of the term because the specifics are not material to the point being made.

These merge semantics effectively make removal akin to a move whose destination is an abstract "removed" location.
This is in tune with the return value of the `treeStatus()` API which will return `TreeStatus.Removed` in that situation.

#### Last Write Wins

It's possible for concurrent edits to represent fundamentally incompatible user intentions.
Whenever that happens, the edit that is sequenced last will win out.

Example 1:
Alice and Bob concurrently change the background color of the same sticky note
such that Alice would change it from yellow to red and Bob would change it from yellow to blue.
If the edits are sequenced such that Alice's edit is applied first and Bob's edit is applied second,
then the background color of the note will change from yellow to red then from red to blue.
If the edits are sequenced in the opposite order,

> then the background color of the note will change from yellow to blue then from blue to red.

Author: 🤯

no matter how complex, and no matter what concurrent edits may have been sequenced before it,
has only one possible effect.
Constrains are a mechanism to allow transactions authors to override this default
so that their transaction will only have that effect if some specific conditions are met,
and will have no effect at all otherwise.
This is useful when the transaction's effect may be rendered undesirable by the effect of concurrent edits that are sequenced before the transaction.

Consider an application whose data model includes two arrays,
with the invariant that the length of one array is expected to always be the same as the length of the other array.
The application allows users to perform the following operations as transactions:

-   Add a new element in each array
-   Remove a single existing element from each array

There's no way for the adding of elements to violate the invariant that the lengths of the two arrays ought to remain the same.
It's possible however for the removal of existing elements to violate this invariant:
consider the starting state `{ arrayA: [1, 2], arrayB: [3, 4] }`.

> Suppose Alice tries to remove element 1 from `arrayA` and element 3 from `arrayB`.

Reviewer 2: ```suggestion
Suppose Alice tries to remove `a1` and `b1`.
```

Concurrently to that, Bob tries to remove element 1 from `arrayA` and element 4 from `arrayB`.
No matter the sequencing order between Alice and Bob's transactions,
the resulting state once both are applied will be `{ arrayA: [2], arrayB: [] }`.

This issue can be addressed by adding a constraint to the transactions that remove elements:
each such transaction can establish a precondition that the nodes to be removed are not already removed.

With such a constraint, the resulting state in the scenario above will depend on the sequencing order:
It will be `{ arrayA: [2], arrayB: [4] }` if Alice's transaction was sequence before Bob's
and `{ arrayA: [2], arrayB: [3] }` otherwise.

### Schema Changes

> At the time of writing,

Reviewer 2: ```suggestion
As of 2024-04-25,
```

all edits/transactions have the implicit constraint that the schema is not changed concurrently to them.
Similarly, all schema changes have the implicit constraint that neither the schema nor the document data is changed concurrently to them.

> This is tolerable because schema changes are rare but will be improved in the future to be less conservative.

Reviewer 2: ```suggestion
This is tolerable because schema changes are rare. The merge semantics will be improved in the future to be less conservative.
```

# SharedTree Merge Semantics

This document offers a high-level description `SharedTree`'s merge semantics.

Target audience: `SharedTree` users and maintainers.

## What Are Merge Semantics?

Merge semantics define how `SharedTree` reconciles concurrent edits.

### Concurrent Edits

When several peers edit the same document, it's possible for some of the edits to be concurrent.

> The edits of two clients are concurrent if they each made those edits before they had received the others’ edits from the server.

Author: I decided to rephrase this.

For example, imagine Alice and Bob are editing a document that contains sticky notes whose background color can be changed.
If Alice changes the background color of one sticky note from yellow to red,
that edit will be sent to the server which will forward it to Bob.

Now imagine Bob wants to change the background of that same sticky note to the color blue.
If Bob makes this change _after_ receiving the edit from Alice, then he will be changing the color from red to blue.

> This is not a case of concurrent editing because Bob had received Alice's edit before changing the color.

Author: The current order has the advantage of putting the simpler examples first, thereby making the example of concurrency more digestable when it finally comes.
I've added "Let's first look at some examples where the edits are not concurrent." so it's less confusing.

On the other hand, if Bob makes this change _before_ receiving the edit from Alice,
then his edit will be concurrent to Alice's edit, and he will be changing the color from yellow to blue.

### Reconciling Concurrent Edits

Reconciling concurrent edits is trivial when they affect independent parts of the tree.

For example, if Alice and Bob concurrently change the background color of _separate_ sticky notes,
then there is only one reasonable outcome.

However, it's possible for some concurrent edits to affect overlapping parts of the tree.
This leads to a situation where there may be multiple reasonable outcomes
(and possibly some unreasonable ones).

For example, if Alice and Bob concurrently change the background color of _the same_ sticky note
such that Alice would change it from yellow to red and Bob would change it from yellow to blue,
then one could imagine multiple possible outcomes:

-   change the color to red
-   change the color to blue
-   keep the color yellow
-   change the color to purple

`SharedTree`'s merge semantics define which outcome will be picked.

## Why/When Should You Care?

> Developers that use `SharedTree` should be able to be productive without constantly worrying about merge semantics.

Author: Saying was is or should be is more informative and direct than saying what isn't or shouldn't be. I've rephrased a bit.

There are, however, situations that warrant an awareness and understanding of merge semantics.
Those are commonly:

-   The need to understand the experience that end-users will face given the host application's usage of `SharedTree`'s editing capabilities.

> -   The need to determine how to use `SharedTree`'s editing capabilities to achieve a desired end-user experience, and whether that experience is achievable in the first place.

Author: I agree it's ver much tied to the first point. I'm not a fan of "mitigate" because it sounds too much like "SharedTree is a pain that you need to deal with". I much prefer the affirmation that it is powerful and you need to know how to weild that power. I also don't want to lose the notion that it may not be possible to do exactly what you want to do. If we don't make it clear that you can't do everything then we run the risk of users getting furstrated looking for a way to do something that's not supported, or thinking we support everything but misinterpretting some outcomes as bugs.

> -   The need to determine how to use `SharedTree`'s editing capabilities to achieve a desired end-user experience, and whether that experience is achievable in the first place.

Reviewer 2: "wield that power" - Is Nick willing to wield such power? Should any of us be? What if such power is too great for us to wield... 

what if it would destroy us?

> -   The need to determine how to use `SharedTree`'s editing capabilities to achieve a desired end-user experience, and whether that experience is achievable in the first place.

Author: It can't - it can remove us, but so long as revertibles are created we won't be destroyed.

The application's editing code may attempt to keep the two arrays' length in sync by always adding to and removing from both array in equal measure.
Despite that, `SharedTree`'s merge semantics are such that a scenario involving concurrent edits may still lead to a state where the arrays end up with different length.
Understanding `SharedTree`'s merge semantics can help the application author anticipate this invariant violation or (failing that) diagnose it after the fact.
Understanding `SharedTree`'s merge semantics will also enable the application author to understand how to remedy this danger,
either by adopting a data model that prevents the issue (e.g., using a single array of pairs),
or by changing the application's editing code (using transactions and constraints) to circumvent it.

## How We Describe Merge Semantics

The merge semantics of a given kind of edit (e.g., moving a node) define what the impact of the edit will be
depending on what concurrent edits were sequenced before it.
This could be captured in a set of rules of the form:
`if <specific-concurrent-edit> was sequenced before this edit then the outcome of this edit will be <specific-outcome>`.

> Capturing the semantics of edits in this way would however be cumbersome because the set of possible concurrent edits and the set possible matching outcome if often open-ended.

Author: I decided to remove this part.

We instead describe the merge semantics of each kind of edit by describing its preconditions and postconditions.

> It is these preconditions and postconditions that define whether an edit's merge semantics make it suitable for a given scenario.

Author: No. I clarified in the other comment thread. Let's chat if that didn't help.

### Preconditions

The preconditions characterize what must be true for the edit to be considered valid.
If those preconditions are not met, then the edit, and the whole transaction it is part of,
is dropped: they have no effect.

For example, imagine Alice inserts a node of type `Foo` in an array,
with the precondition that by the time the edit is finally applied,
this array allow must instances of type `Foo`.
Concurrently to that, Bob edits the document schema such that the array in question no longer allows instances of type `Foo`,
with the precondition that by the time the edit is finally applied,
no instances of type `Foo` must exist in the array.
If Alice's edit is sequenced first, then the node will successfully be inserted, and Bob's edit will be dropped.
If Bob's edit is sequenced first, then the schema will successfully be changed, and Alice's edit will be dropped.

> When considering whether a given edit is suitable for a given scenario,

Author: No. Preconditions determine whether an edit is valid in a given situation or not. This sentence is about the suitability of an edit for a given scenario given what those preconditions are.

> When considering whether a given edit is suitable for a given scenario,

Author: If somethign is invalid then it will not be allowed to happen. By contrast, if something is not suitable, it will be allowed to happen but there was possbily something better to do.

For example, the postcondition for our implementation of move
(that the moved node will reside at the chosen destination)
is not predicated on the node not having been concurrently removed or moved.
This means that when it comes to the final location of the node that is targeted by the move,
there is only one possible outcome
(the node being at the chosen destination)
so long as the preconditions of the move were met.

Note that such conditional postconditions would be different from additional preconditions.
For example, adding a precondition that the node moved node must not have been concurrently removed would,
in a situation where that did occur,
render such a move _and the whole transactions it figures in_ invalid.
By contrast, making the postcondition conditional on the node not having been concurrently removed would simply render the move ineffective,
but not adversely affect the rest of the transaction.

As with minimal preconditions, we may support more variations of our current set of edits in the future.

This prioritization is partially guided by the same motivation that guided our choice with minimal preconditions:

> it makes more for a greater set of possible edits when combined with [constraints](#constraints).

Reviewer: Grammar

4. The client sends the edit to the service.
5. The service orders the edit relative to other (possibly concurrent) edits based on arrival time.
6. The service broadcasts the edit to all connected clients.
7. Each client receives the edit and reconciles it with concurrent edits that were sequenced before.
   This is where the merge semantics of the edits to be reconciled are taken into account.
8. Each client applies the appropriate effects of the edit (if any).

Steps 7 and 8 are somewhat different for the client that initiated the edit:
since it already knows of the edit and already applied it in step 3,
it only needs to adjust its local state to reflect the impact of concurrent edits that were sequenced before the edit it initiated.

### A Conflict-Free Model

`SharedTree` never models edits (or documents) as being "conflicted".
Indeed, `SharedTree` never surfaces a conflict where one of two states or one of two edits must be selected as the winner or manually reconciled.
(`SharedTree` may offer capabilities to facilitate the creation of such end user experiences in the future,
but these would only exist atop `SharedTree`'s core conflict-free model.)

> In that regard, `SharedTree` closer to operational transform or CRDT systems, and different from systems like [Git](https://git-scm.com/).

Reviewer: ```suggestion
In that regard, `SharedTree` is closer to operational transform or CRDT systems, and different from systems like [Git](https://git-scm.com/).
```

\*[Constraints](#constraints) is a feature that allows transaction authors to add additional preconditions to a transaction.
This makes understanding preconditions a prerequisite for using constraints,
but constraints can be ignored when it comes to understanding how preconditions affect merge semantics in general.

Note that the preconditions of a transaction are the union of the preconditions of its edits.

### Postconditions

The postconditions characterize what is guaranteed to be true about the effect of the edit provided the preconditions were satisfied.
The term "intention" is often used in relation to an edit's postconditions: they capture what the purpose of the edit is.
This is often expressed in terms of the resulting state after the edit is applied.
For example, move edits guarantee that the targeted nodes end up at the specified destination.
In other words, the intention for a move edit is to have the targeted node reside at the destination.

When expressed in terms of the document state,
these postconditions typically only hold immediately after the edit as opposed to holding indefinitely.

> In our move example, edits that come after the move

Author: Yeah, it was not specific enough to warrant that language.

(whether they were made concurrently or not)
may very well cause those nodes to end up in a different location.

When considering whether a given edit is suitable for a given scenario,
the postconditions help answer the following questions:

1.  Will this edit, in scenarios where it applies, have all the effects I want it to have?
2.  Will this edit, in scenarios where it applies, have none of the effects I do not want it to have?

> The answer to the first question will be negative if some desirable postcondition is missing or an existing one is too narrow.

Reviewer 3: Interesting that this is the only one of the four questions where you don't specify if it can be remedied. Did you want to mention "this is where we add more operations such as slice" or something? :)

this is true even when removals are involved:
if the removal is sequenced last, then the node will be moved and then removed.
If the move is sequenced last, then the node will be removed and then moved.

### Non-Disjunctive Postconditions

Disjunctive postconditions are postconditions of the form `<A> or <B> or ...`.
None of `SharedTree`'s currently supported edits are of this form,
which means they each have a single effect as opposed to one of many effects.

For example, the postcondition for our implementation of move
(that the moved node will reside at the chosen destination)
is not predicated on the node not having been concurrently removed or moved.
This means that when it comes to the final location of the node that is targeted by the move,
there is only one possible outcome
(the node being at the chosen destination)
so long as the preconditions of the move were met.

Note that such conditional postconditions would be different from additional preconditions.
For example, adding a precondition that the moved node must not have been concurrently removed would,
in a situation where that did occur,
render such a move _and the whole transactions it figures in_ invalid.

> By contrast, making the postcondition conditional on the node not having been concurrently removed would simply render the move ineffective

Reviewer 3: You introduce a definition of disjunctive postconditions above but never use it again in the examples. Perhaps illustrating how this example maps to the above A | B | ... form would make that clearer.

By the time this transaction is sequenced and applied,
concurrent edits may have affected the relevant sticky notes in different ways:
some of them may have been assigned different ordinals,
some of them may have been moved/removed,
and some of their parent pages may have been moved/removed.
Despite that, `SharedTree`'s current merge semantics guarantee that by the end of the transaction,
all of the relevant sticky notes will reside in the new page and that their ordinals will be assigned in order from 1 to N.
If any of the concurrent changes had the power to prevent the relevant notes from being moved by the transaction or from being annotated with the ordinals,
then the transaction may lead to a state where the new page only contains a subset of the selected notes, and their ordinals may have gaps, may not be unique, or may be out of order.
In other words, it would be very difficult to make any kind of valuable claim about the effect of the transaction,
and end users would experience some very confusing outcomes.

To get a more general sense of the positive effect of simpler postconditions,
we can model transactions as being composed of `N` edits (`e1` through `eN`),

> where the effect of each edit `ei` is one out of a set whose size is `ki`.

Reviewer 3: ```suggestion
where the effect of each edit `ei` is one out of a set whose size is `ki` where `ki` is the number of postconditions for `ei`.
```

Author: Postconditions are always all true, so having multiple postcondions is not the problem. The probem is having a single (or many, but that's not material here) postcondition that is itself disjunctive.
This making me realize I need to make this clearer at the start of the section. My use of the plural is confusing.

The particular effect that is applied (out of the `ki` possibilities) for a given `ei` depends on what concurrent edits were sequenced before the transaction.
Under this model, the number of possible effects for a transaction of `N` edits is `k1 * k2 * ... * kN`.
`SharedTree`'s supported set of edits all have exactly one possible effect, so any transaction is guaranteed to have `1 * 1 * ... * 1 = 1` effect.
If `SharedTree`'s supported set of edits had not one but two possible effects,
then a transaction of `N` edits would have `2`<sup>`N`</sup> possible effects,
which would make using those edits in transactions less practical.

> ## Constraints

Reviewer 3: Similar to the schema changes section, I'd note that as of now we only have one constraint.
